R/corridor.R

```diff
@@ -97,40 +104,72 @@ river_buffer <- function(river, buffer, bbox = NULL) {
+#' The end points are selected as the most external river crossing edges of the
+#' network within the area of interest. Note that if the river intersects the
+#' area of interest multiple times, only the longest intersecting segment is
+#' considered here.
 #'
 #' @param river A simple feature geometry representing the river centerline
-#' @param bbox Bounding box defining the extent of the area of interest
+#' @param network The spatial network used for the delineation, either provided
+#'   as a [`sfnetworks::sfnetwork`] or as a [`sf::sf`]/[`sf::sfc`] object (edges
+#'   only)
+#' @param aoi Area of interest, provided as a bounding box or as polygon. If not
+#'   given, it is taken as the bounding box of the spatial network
 #'
 #' @return A simple feature geometry including a pair of points
-corridor_end_points <- function(river, bbox) {
-  aoi <- sf::st_as_sfc(bbox)
-  aoi_boundary <- sf::st_boundary(aoi)
-  intersections <- sf::st_intersection(river, aoi_boundary)
-  sf::st_cast(intersections, "POINT")
+corridor_end_points <- function(river, network, aoi = NULL) {
+  # Only consider the edges of the spatial network
+  spat_net_edges <- sf::st_geometry(sf::st_as_sf(network, "edges"))
+
+  # Set area of interest as a polygon, even if not defined
+  if (is.null(aoi)) aoi <- sf::st_bbox(spat_net_edges)
+  aoi <- as_sfc(aoi)
+
+  # Clip the river geometry using the area of interest. This might lead to
+  # multiple river segments
+  river_segments <- sf::st_intersection(aoi, sf::st_cast(river, "LINESTRING"))
+
+  # Build a spatial network using the river segments and select the connected
+  # component with overall longest edge length
+  river_network <- as_network(river_segments, flatten = FALSE, clean = FALSE)
+  components <- tidygraph::morph(river_network, tidygraph::to_components)
+  sum_edge_lengths <- function(x) sum(sf::st_length(sf::st_as_sf(x, "edges")))
+  edge_lengths <- sapply(components, sum_edge_lengths)
+  river_network <- components[[which.max(edge_lengths)]]
+  river_network <- clean_network(river_network)
+
+  # Find intersections with the spatial network, and add these points as nodes
+  # to the river network. These nodes are the candidates corridor end points
+  river_network_edges <- sf::st_geometry(sf::st_as_sf(river_network, "edges"))
+  intersections <- sf::st_intersection(river_network_edges, spat_net_edges)
+  river_network <- sfnetworks::st_network_blend(river_network, intersections)
+  nodes <- nearest_node(river_network, intersections)
+
+  # Pick a reference point on the selected river segment, either up- or down-
+  # stream from all crossings. Pick this as the closest node to the aoi boundary
+  reference <- nearest_node(river_network, sf::st_boundary(aoi))
+
+  # Compute the network distance from the reference node to all candidate end
+  # points. The closest and furthest nodes are selected as corridor end points
+  river_network <- add_weights(river_network, weight_name = "weight")
```

Comment:
```suggestion
  river_network <- add_weights(river_network)
```
This is the default and thus not necessary, right? Unless we use `weights = weight_name` in `sfnetworks::st_network_cost()` below and want to make the link explicit?

Reply:
Correct, it is not necessary, but I indeed wanted to make the link explicit because `sfnetworks::st_network_cost` neglects weights if the `weights` input argument is not specified..